### PR TITLE
[GLUTEN-2038][VL] Improve parquet write using arrow chunked array

### PR DESCRIPTION
**Pull request by ulysses-you**
## What changes were proposed in this pull request?

Velox parquet write would build row groups for each columnar batch, which causes a lot of small row groups. We can cache columnar batch to make a bigger one.

This pr adds a new option `parquet.block.rows` to control the numRows of one row group.

The velox related code change see https://github.com/oap-project/velox/pull/326

(Fixes: \#2038)

## How was this patch tested?
improve test and manually test:

```
spark.range(1000000).write.format("velox").mode("overwrite").option("compression","zstd").save("/opt/parquet/test")

java -cp 'parquet-tools-1.11.1.jar:spark-3.3.1/jars/*' org.apache.parquet.tools.Main meta /opt/parquet/test/part-00000-478ee552-27da-4dc9-9168-6af5303a983d-c000.zstd.parquet

file schema: schema
--------------------------------------------------------------------------------
id:          REQUIRED INT64 R:0 D:0

row group 1: RC:1000000 TS:8297934 OFFSET:4
--------------------------------------------------------------------------------
id:           INT64 ZSTD DO:4 FPO:134785 SZ:1299824/8297934/6.38 VC:1000000 ENC:PLAIN,RLE_DICTIONARY,RLE ST:[min: 0, max: 999999, num_nulls: 0]
```

```
spark.range(1000000).write.format("velox").mode("overwrite").option("compression","zstd").option("parquet.block.size","3000000").save("/opt/parquet/test")

file schema: schema
--------------------------------------------------------------------------------
id:          OPTIONAL INT64 R:0 D:1

row group 1: RC:300000 TS:2697618 OFFSET:4
--------------------------------------------------------------------------------
id:           INT64 ZSTD DO:4 FPO:134785 SZ:588644/2697618/4.58 VC:300000 ENC:RLE,RLE_DICTIONARY,PLAIN ST:[min: 0, max: 299999, num_nulls: 0]

row group 2: RC:300000 TS:2697618 OFFSET:134123
--------------------------------------------------------------------------------
id:           INT64 ZSTD DO:0 FPO:134123 SZ:587996/2697618/4.59 VC:300000 ENC:RLE,RLE_DICTIONARY,PLAIN ST:[min: 300000, max: 599999, num_nulls: 0]

row group 3: RC:300000 TS:2697618 OFFSET:134123
--------------------------------------------------------------------------------
id:           INT64 ZSTD DO:0 FPO:134123 SZ:587986/2697618/4.59 VC:300000 ENC:RLE,RLE_DICTIONARY,PLAIN ST:[min: 600000, max: 899999, num_nulls: 0]

row group 4: RC:100000 TS:1012795 OFFSET:101594
--------------------------------------------------------------------------------
id:           INT64 ZSTD DO:0 FPO:101594 SZ:308849/1012795/3.28 VC:100000 ENC:RLE,RLE_DICTIONARY,PLAIN ST:[min: 900000, max: 999999, num_nulls: 0]
```

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @JkSelf @zhouyuan , I used arrow parquet writer directly rather than velox parquet writer. Is there any concerns ?

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala [hunk 1]
```diff
@@ -62,12 +62,15 @@ class VeloxParquetFileFormat extends GlutenParquetFileFormat
     val parquetOptions = new ParquetOptions(options, sparkSession.sessionState.conf)
     conf.set(ParquetOutputFormat.COMPRESSION, parquetOptions.compressionCodecClassName)
     // pass options to native so that velox can take user-specified conf to write parquet,
-    // i.e., compression and block size.
+    // i.e., compression, block size, block rows.
```
Comment (Yohahaha) [inline backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala:65]: Do we need get this config from Spark/Hadoop's configuration? I think it would be better to decided by Spark/Hadoop, instead of Gluten.

Comment (ulysses-you) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala:65]: I agree, will add two configs in gluten config.

**Comment by JkSelf** [comment]
> cc @JkSelf @zhouyuan , I used arrow parquet writer directly rather than velox parquet writer. Is there any concerns ?

We are currently removing the velox to arrow conversion in gluten. So it's better to use the velox parquet writer.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala (hunk 1 above)
Comment (Yohahaha) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala]: What I mean is to use pre-defined config from parquet, instead of Gluten defined, cause Gluten should not affect parquet behavior, it should belong to Spark.

Consider this UT from Spark https://github.com/apache/spark/blob/a8ea35f7c2fbd80c3902707ecf224d50800ba622/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilterSuite.scala#L1933-L1939

And I could not find parquet.block.rows in parquet-mr project, could you clarify for this config?

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 2]
```diff
@@ -725,6 +732,18 @@ object GlutenConfig {
...
+      .internal()
+      .longConf
+      .createWithDefault(128 * 1024 * 1024)
+
+  val COLUMNAR_PARQUET_WRITE_BLOCK_ROWS =
+    buildConf("spark.gluten.sql.native.parquet.write.blockRows")
+      .internal()
+      .longConf
+      .createWithDefault(100 * 1000 * 1000)
+
```
Comment (Yohahaha) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:746]: Consider using `ParquetOutputFormat.BLOCK_SIZE`

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala [hunk 3]
```diff
@@ -235,6 +241,7 @@ object GlutenConfig {
   val SPARK_HIVE_EXEC_ORC_COMPRESS: String = SPARK_PREFIX + HIVE_EXEC_ORC_COMPRESS
   val SPARK_SQL_PARQUET_COMPRESSION_CODEC: String = "spark.sql.parquet.compression.codec"
   val PARQUET_BLOCK_SIZE: String = "parquet.block.size"
+  val PARQUET_BLOCK_ROWS: String = "parquet.block.rows"
```
Comment (Yohahaha) [inline shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:244]: Could you clarify where `parquet.block.rows` from? I could not find it in parquet-mr project.

File: backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala (hunk 1 above)
Comment (ulysses-you) [reply backends-velox/src/main/scala/org/apache/spark/sql/execution/datasources/velox/VeloxParquetFileFormat.scala]: Parquet or Spark have no config `parquet.block.rows`. This config is used to align with arrow parquet write which has the concept, see `WriterProperties.max_row_group_length`.

File: shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala (hunk 2 above)
Comment (ulysses-you) [reply shims/common/src/main/scala/io/glutenproject/GlutenConfig.scala:746]: `ParquetOutputFormat.BLOCK_SIZE` can not been propagated by Spark SQLConf, so I add two more configs to help user change at runtime.

File: backends-velox/src/test/scala/io/glutenproject/execution/VeloxParquetWriteSuite.scala [hunk 4]
```diff
@@ -37,19 +37,21 @@ class VeloxParquetWriteSuite extends WholeStageTransformerSuite {
...
 
-        withTempPath { f =>
-          spark.table("lineitem").write
-            .format("velox")
-            .option("compression", codec)
-            .save(f.getCanonicalPath)
-          val files = f.list()
-          assert(files.nonEmpty, extension)
-          assert(files.exists(_.contains(extension)), extension)
+        TPCHTables.foreach { case (_, df) =>
```
Comment (JkSelf) [inline backends-velox/src/test/scala/io/glutenproject/execution/VeloxParquetWriteSuite.scala:40]: Can you help to delete the unit test in [TestOperator ](https://github.com/oap-project/gluten/blob/8f2c139700c3e253a65c4b2158a9a00809e1ddfd/backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala#L95)here? It seems this test contains all the parquet write case. Thanks.

Comment (ulysses-you) [reply backends-velox/src/test/scala/io/glutenproject/execution/VeloxParquetWriteSuite.scala:40]: sure, cleanup that

File: ep/build-velox/src/get_velox.sh [hunk 5]
```diff
@@ -2,8 +2,8 @@
 
 set -exu
 
-VELOX_REPO=https://github.com/oap-project/velox.git
-VELOX_BRANCH=main
+VELOX_REPO=https://github.com/ulysses-you/velox.git
```
Comment (JkSelf) [inline ep/build-velox/src/get_velox.sh:5]: Please help to revert this change. Thanks.

Comment (ulysses-you) [reply ep/build-velox/src/get_velox.sh]: reverted

**Review by JkSelf — APPROVED** [review]
